<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>lassosum · PRS.jl</title><link rel="canonical" href="https://mfz.github.io/PRS.jl/lassosum/"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">PRS.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>lassosum</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>lassosum</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/mfz/PRS.jl/blob/master/docs/src/lassosum.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="lassosum"><a class="docs-heading-anchor" href="#lassosum">lassosum</a><a id="lassosum-1"></a><a class="docs-heading-anchor-permalink" href="#lassosum" title="Permalink"></a></h1><p>Mak et al.(2018), Polygenic scores via penalized regression on summary statistics</p><p>Consider linear regression (GWAS) problem</p><p class="math-container">\[\mathbf{y} = \mathbf{X\beta} + \mathbf{\epsilon}\]</p><p>where <span>$\mathbf{X}$</span> is the n x p genotype matrix (normalized to sd(cols) = 1) and <span>$\mathbf{y}$</span> is n x 1 phenotype matrix (normalized to sd(cols) = 1)</p><p>Using regularization, we arrive at the LASSO with objective function</p><p class="math-container">\[f(\mathbf{\beta}) = (\mathbf{y} - \mathbf{X\beta})^T (\mathbf{y} - \mathbf{X\beta}) + 2 \lambda ||\beta||_1\]</p><p class="math-container">\[f(\mathbf{\beta}) = \mathbf{y}^T \mathbf{y} - 2 \mathbf{\beta}^T \mathbf{X}^T \mathbf{y} + \mathbf{\beta}^T \mathbf{X}^T \mathbf{X} \mathbf{\beta} + 2 \lambda ||\beta||_1\]</p><p>Observing that </p><p class="math-container">\[\mathbf{r} = \mathbf{X}^T \mathbf{y}\]</p><p>is the genotype - phenotype correlation (obtained from p-value in summary statistics) and</p><p class="math-container">\[\mathbf{R} = \mathbf{X}^T \mathbf{X}\]</p><p>is the LD matrix, </p><p>the objective can be rewritten as</p><p class="math-container">\[f(\mathbf{\beta}) = \mathbf{y}^T \mathbf{y} - 2 \mathbf{\beta}^T \mathbf{r} + \mathbf{\beta}^T \mathbf{R} \mathbf{\beta} + 2 \lambda ||\beta||_1\]</p><p>The LD matrix <span>$\mathbf{R}$</span> could also be derived from reference data. In this case, <span>$\mathbf{r}$</span> and <span>$\mathbf{R}$</span> are not derived from the same genotype matrix <span>$\mathbf{X}$</span> anymore, implying that the objective function is no longer a penalized least-squares (or LASSO) problem.</p><p>It can be turned into a LASSO problem when regularization</p><p class="math-container">\[\mathbf{R_s} = (1-s)\mathbf{R} + s\mathbf{I}, 0 \le s \le 1\]</p><p>is applied (see simple proof in paper). The shrinkage parameter <span>$s$</span> determines how much of the linkage disequilibrium between different variants is taken into account.</p><p>This renders the objective function an elastic net,</p><p class="math-container">\[f(\mathbf{\beta}) = \mathbf{y}^T \mathbf{y} - 2 \mathbf{\beta}^T \mathbf{r} + (1-s)\mathbf{\beta}^T \mathbf{R} \mathbf{\beta} + s \mathbf{\beta}^T \mathbf{\beta} + 2 \lambda ||\beta||_1\]</p><p>,</p><p>which can be solved using coordinate descent.</p><p>In each iteration of coordiante descent, the contribution of each single coordinate is optimized on its own. For a single coordinate <span>$\beta_l$</span>, the contributions to <span>$f(\beta)$</span> are</p><p class="math-container">\[\beta_l^2 (s + (1-s)\mathbf{R}_{ll}) - 2 \beta_l (r_l - (1-s)\mathbf{R}_{l}\beta + (1-s)\mathbf{R}_{ll}\beta_l) + 2\lambda |\beta_l|\]</p><p>NOTE: Is not <span>$s + (1-s)\mathbf{R}_{ll} = 1$</span>? Only if <span>$R_{ll}$</span> = 1, but for some markers the variance is 0, hence <span>$R_{ll}$</span> = 0</p><p>Defining </p><p class="math-container">\[u_l = r_l - (1-s)R_{.l} \beta + (1-s) R_{ll} \beta_l\]</p><p>the minimum contribution is obtained at</p><p class="math-container">\[\beta_l = \mathmr{sgn}(u_l) (|u_l| - \lambda) / (s + (1-s)\mathbf{R}_{ll}) , |u_l| - \lambda &gt; 0\]</p><p class="math-container">\[\beta_l = 0, \mathrm{otherwise}\]</p><p>NOTE: Is this not the same as is done in LDpred? I.e. we look at the contribution of all SNPs except <span>$l$</span> to <span>$\mathbf{r}_l$</span> due to LD, and then apply some shrinkage to <span>$\beta_l$</span>.</p><p>NOTE: Implementation where thresholded LD matrix is used suffers from divergence issues. Implementation using genotype matrix directly converges fine.</p></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 5 January 2021 17:37">Tuesday 5 January 2021</span>. Using Julia version 1.5.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
